linux/mcompile/keymap/keymap.cpp

```diff
@@ -109,355 +472,330 @@ void Split_US_To_3D_Vector(v_str_3D &all_US,v_str_1D completeList) {
-  std::string aaaa = get_Other_Char_FromUS( "Y", All_Vector);
-
-  std::string b = get_US_Char_FromOther( "z", All_Vector);
-  std::string bb = get_US_Char_FromOther( "Z", All_Vector);
-  std::string bbb = get_US_Char_FromOther( "y", All_Vector);
-  std::string bbbb = get_US_Char_FromOther( "Y", All_Vector);
-
-  std::string c = getKeyNrOf_OtherChar( "z", All_Vector);
-  std::string cc = getKeyNrOf_OtherChar( "Z", All_Vector);
-  std::string ccc = getKeyNrOf_OtherChar( "y", All_Vector);
-  std::string cccc = getKeyNrOf_OtherChar( "Y", All_Vector);
-
-  std::string d = getKeyNrOf_USChar( "z", All_Vector);
-  std::string dd = getKeyNrOf_USChar( "Z", All_Vector);
-  std::string ddd = getKeyNrOf_USChar( "y", All_Vector);
-  std::string dddd = getKeyNrOf_USChar( "Y", All_Vector);
-
-  std::cout << "get_Other_Char_FromUS z-Z-y-Y: "  << ".." << a<< ".." <<aa<< ".." <<aaa<< ".." <<aaaa<< ".." << "\n";
-  std::cout << "get_US_Char_FromOther z-Z-y-Y: "  << ".." << b<< ".." <<bb<< ".." <<bbb<< ".." <<bbbb<< ".." << "\n";
-  std::cout << "getKeyNrOf_OtherChar z-Z-y-Y: "   << ".." << c<< ".." <<cc<< ".." <<ccc<< ".." <<cccc<< ".." << "\n";
-  std::cout << "getKeyNrOf_USChar z-Z-y-Y: "      << ".." << d<< ".." <<dd<< ".." <<ddd<< ".." <<dddd<< ".." << "\n";
-}
+  //ALT-GR + CAPS (shiftstate: 6)
+  else if (( ss == MenuCtrl ) && ( caps == 1 )){
+    GdkModifierType MOD_AltGr = (GdkModifierType) (GDK_MOD2_MASK | GDK_MOD5_MASK | GDK_LOCK_MASK);
+    gdk_keymap_translate_keyboard_state (keymap, keycode, MOD_AltGr , 0, keyvals, NULL, NULL, & consumed);
+  }
 
-void print_simple_map_US(v_str_3D &All_Vector, int shiftstate){
-  std::string out,diff;
-  // ? CHECK if ran OK-> return 0/1
-  printf("-----------------------------------------------------------------------------------------------------------------------------------------------\n");
-  for ( int i=0; i< (int)All_Vector[0].size();i++) {
-    out =get_Other_Char_FromUS(All_Vector[0][i][shiftstate], All_Vector);
+  //ALT-GR (shiftstate: 7)
+  else if (( ss == ShftMenuCtrl ) && ( caps == 0 )){
+    GdkModifierType MOD_AltGr = (GdkModifierType) ( (GDK_SHIFT_MASK | GDK_MOD2_MASK | GDK_MOD5_MASK) );
+    gdk_keymap_translate_keyboard_state (keymap, keycode, MOD_AltGr , 0, keyvals, NULL, NULL, & consumed);
   }
+
+  //ALT-GR +CAPS (shiftstate: 7)
+  else if (( ss == ShftMenuCtrl ) && ( caps == 1 )){
+    GdkModifierType MOD_AltGr = (GdkModifierType) ( (GDK_SHIFT_MASK | GDK_MOD2_MASK | GDK_MOD5_MASK | GDK_LOCK_MASK) );
+    gdk_keymap_translate_keyboard_state (keymap, keycode, MOD_AltGr , 0, keyvals, NULL, NULL, & consumed);
+  }
+  else
+    return 0;
+
+  return (int) *keyvals;
+}
+
+KMX_DWORD KMX_get_KeyValUnderlying_From_KeyCodeUnderlying(GdkKeymap *keymap, guint keycode, int shift_state_pos) {
+  GdkKeymapKey *maps;
+  guint *keyvals;
+  gint count;
+  KMX_DWORD KVal;
+
+  if (!gdk_keymap_get_entries_for_keycode(keymap, keycode, &maps, &keyvals, &count))
+    return 0;
+
+  if (!(shift_state_pos <= count))
```

Comment:
Isn't this off by one, i.e. shouldn't it be `if (!(shift_state_pos < count))` ? And it's easier to read if you remove the negation:

```suggestion
  if (shift_state_pos >= count)
```

Reply:
OK-done -> PR [#11886](https://github.com/keymanapp/keyman/pull/11886):
off by one: you were right

shift_state_pos ( now renamed to gdk_level) is the level for GDK : 
Level 0 = Base; 
Level 1 = Shift; 
Level 2 = ALT; 
Level 3 = ALT+ Shift; 

Before we use the function gdk_translate_keyboard_state() we make sure that the level exists e.g. if for a keyboard count = 16, level can maximal be 15. This is what !(shift_state_pos < count) ( now moved to ensureValidInputForKeyboardTranslation() ) checks.